### DISCUSSION TITLE
Share your deployment strategies!

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#763] [General]
We are establishing a rapid release cycle for our oqtane enterprise app.   So far, the approach I've been taking is to install Oqtane on IIS following the guidance, and then deploying the themes/modules by packaging them up and installing them.  This feels like the right way to deploy, but I'm wondering if folks have different strategies.

Before Oqtane, we'd deploy our apps was using visuals studio's Publish feature as well as the Publish tasks in CI. I'm courious if folks are using that approach.

Another approach I'm working toward is an in-docker build of the app and deploy to K8s.  This would have a dockerfile that would build up each layer starting with oqtane, adding themes and modules, then config, resulting in a deployable image.  This is a little more ambitious, but reasonable.  I'm curious if folks are exploring that path.

Any other paths to deploy to dev/test as well as production would be great to hear about.

**Comment by sbwalker** [comment]
Visual Studio's publish capability will not work with the Oqtane application itsself because just like every other Microsoft solution Visual Studio relies on static, hard-coded references and dependencies. Oqtane is dynamic so it does not include static references to external modules or themes. As a result, if you are using modules or themes in development and attempt to do a Publish, Visual Studio will not deploy the module and theme assemblies.

**Reply by markdav-is** [reply]
ya, Publish seems to be a bit of a black box when it comes to deploying.  I have not dug into it much to see if you can hook in more build steps.  It would be nice to start with the base Oqtane, add the themes, modules, content and then publish.  I'll look at AzureDevops that has more steps in the build pipeline and see what's possible there.

I'm mostly just fishing for stories... how are folks deploying this today with themes and modules attached?

**Reply by mikecasas** [reply]
@markdav-is Could you have the themes and modules on an image running a web server? Maybe a private nuget container/server, then somehow auto install the modules and themes?

**Reply by markdav-is** [reply]
That's one way to publish a docker images.  You can build it up from a base and then use that as a base for your instances.  We used one of your base images that way a few months back.  I'd like to see the whole layer-cake be baked from source, but that's more of a long term goal.  If themes and modules are all nuget packages we might be able to start with a base images and then add those packages inside the docker following something similar to this: https://stackoverflow.com/questions/48642947/how-to-add-nuget-packages-to-the-dockerfile-build-steps

**Comment by louyo** [comment]
Being the project dunce, I use publish. I publish via FTP to a Linux system, then install the modules. I have no custom themes at the moment. My use case is very small and probably out in left field somewhere.
